Move the tiling logic out of DecomposePackUnPackOps pass.

hanhanW opened the pull request:
It is not necessary to make outer dims be all unit dims before decomposition. It exposes the tiling sizes to configuration, and will unblock other codegen paths in IREE, e.g., specialized transpose lowering on AVX2 and AVX512, etc.

The lowerPack and lowerUnPack patterns are required for fusion cases. The outer dims could be non unit in fusion because the tiling sizes are driven by other Linalg ops. In this context, it relies on lowerPack and lowerUnpack to decompose the tensor.pack/unpack ops. It can also be used in specialized transpose lowering.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -585,6 +585,8 @@ void addMmt4dTilingExpertPassPipeline(OpPassManager &passManager,
 void addCPUDataTilingPipeline(OpPassManager &passManager) {
   addTileAndDistributePasses(passManager);
   OpPassManager &nestedModulePM = passManager.nest<ModuleOp>();
+  nestedModulePM.addNestedPass<func::FuncOp>(
```
This will affect all MMT4D ops?

hanhanW replied:
No, it won't affect mmt4d ops. The mmt4d ops go with `Mmt4dTilingExpertPassPipeline`

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp:
```diff
@@ -41,6 +75,23 @@ struct DecomposePackUnPackOpsPass
+    patterns.add<linalg::GeneralizeOuterUnitDimsPackOpPattern,
+                 linalg::GeneralizeOuterUnitDimsUnPackOpPattern>(
+        ctx,
+        /*benefit=*/10);
```
Just pointing out, but not sure I have a suggestion here. Using pattern benefit is not a very stable way of making sure patterns apply in an order. Maybe have two separate `applyPatternsAndFoldGreedily` methods to have one trigger after another.

hanhanW replied:
I thought by design they are expected to apply patterns in an order if we assign benefits for all the patterns? The driver processes ops in a worklist-driven fashion and greedily applies the patterns that locally have the most benefit, while the benefits are assigned by us?

https://mlir.llvm.org/docs/PatternRewriter/#greedy-pattern-rewrite-driver

MaheshRavishankar replied:
That's the theory. It's not a very robust system. It's ok in this case (that's why I didn't push it) but better to not rely on ot

hanhanW replied:
I see, thanks!

hanhanW committed:
address comments

hanhanW committed:
Move the tiling logic out of DecomposePackUnPackOps pass.

It is not necessary to make outer dims be all unit dims before
decomposition. It exposes the tiling sizes to configuration, and will
unblock other codegen paths in IREE, e.g., specialized transpose
lowering on AVX2 and AVX512, etc.

The lowerPack and lowerUnPack patterns are required for fusion cases.
The outer dims could be non unit in fusion because the tiling sizes are
driven by other Linalg ops. In this context, it relies on lowerPack and
lowerUnpack to decompose the tensor.pack/unpack ops. It can also be used
in specialized transpose lowering.
